File: refactor/Parser.py
import Expr
import Type

endl = "\n"
FLAG_DEFAULT = 0
FLAG_STRING = 1
FLAG_ESCAPING_STRING = 2
FLAG_COMMENT = 3 
ESCAPING_LIST = {
    "n": "\n",
    '"': "\"",
    "\\": "\\",
    "0": "\0",
}

def is_quote_by(s, q):
    return len(s) > 1 and q == s[0] == s[-1]

def parse_value(s):
    if is_quote_by(s, '"'):
        return Expr.Value(String(''.join(s[1:-1])))
    else:
        s = ''.join(s)
        if s.startswith("0x"):
            return Expr.Value(int(s, 16))
        elif s.startswith("0b"):
            return Expr.Value(int(s, 16))
        elif s.startswith("0") and "." not in s:
            return Expr.Value(int(s, 8))
        try:
            return Expr.Value(int(s))
        except Exception: pass
        try:
            return Expr.Value(float(s))
        except Exception: pass
    return Expr.Symbol(val)

def escape(char):
    if char in ESCAPING_LIST:
        return ESCAPING_LIST[char]
    else:
        raise SyntaxError(f"{char} can't be escaped")

def parse(expr):
    res: Expr.Expression = Expr.Expression()
    last = [res]
    buffer = []
    FLAG = 0
    lineNum = 1
    charNum = 0
    index = -1
    length = len(expr)
    while index < length - 1:
        charNum += 1
        index += 1
        char = expr[index]
        # print("char:", char)
        # print("res:", res)
        # print("last:", last)
        # print("buffer:", buffer)
        # print("FLAG:", FLAG)
        if FLAG == FLAG_COMMENT:
            if char == "\n":
                FLAG = FLAG_DEFAULT
        elif FLAG == FLAG_ESCAPING_STRING:
            if char == "x":
                buffer.append(chr(int(expr[index + 1:index + 3], 16)))
                index += 2
                FLAG = FLAG_STRING
            elif char == "u":
                buffer.append(chr(int(expr[index + 1:index + 5], 16)))
                index += 4
                FLAG = FLAG_STRING
            else:
                buffer.append(escape(char))
                FLAG = FLAG_STRING
        elif FLAG == FLAG_STRING:
            if char == "\\":
                FLAG = FLAG_ESCAPING_STRING
            elif char == "\"":
                FLAG = FLAG_DEFAULT
                buffer.append('"')
            else:
                buffer.append(char)
        elif char == "(" or char == "[":
            if buffer:
                last[-1].append(parse_value(buffer))
                buffer = []
            if is_quote(last[-1]):
                new = Type.Quote()
            else:
                new = Expr.Expression()
            last[-1].append(new)
            last.append(new)
        elif char == ")" or char == "]":
            l = last.pop()
            if buffer:
                l.append(parse_value(buffer))
                buffer = []
        elif char == " " or char == "\n" or char == "\t":
            if char == "\n":
                lineNum += 1
                charNum = 0
            if buffer:
                last[-1].append(parse_value(buffer))
                buffer = []
        elif char == "'":
            if index + 1 >= length or\
               (expr[index + 1] != "(" and\
               expr[index + 1] != "["):
                raise SyntaxError(f"""{expr.split(endl)[lineNum-1]}\n{'-' * (charNum - 1 + 13)}^""")
            index += 1
            new = Quote()
            last[-1].append(new)
            last.append(new)
        elif char == ",":
            if not is_quote(last[-1]):
                raise SyntaxError(f"""{expr.split(endl)[lineNum-1]}\n{'-' * (charNum - 1 + 13)}^""")
            if index + 1 >= length or\
               (expr[index + 1] != "(" and\
               expr[index + 1] != "["):
                raise SyntaxError(f"""{expr.split(endl)[lineNum-1]}\n{'-' * (charNum - 1 + 13)}^""")
            index += 1
            new = Expr.Expression()
            last[-1].append(new)
            last.append(new)
        elif char == "\"":
            if FLAG == FLAG_DEFAULT:
                FLAG = FLAG_STRING
                buffer.append('"')
            else:
                # 估計是沒完結的字串..
                raise SyntaxError(f"""{expr.split(endl)[lineNum-1]}\n{'-' * (charNum - 1 + 13)}^""")
        elif char == "\\":
            if FLAG == FLAG_DEFAULT:
                raise SyntaxError(f"""{expr.split(endl)[lineNum-1]}\n{'-' * (charNum - 1 + 13)}^""")
            else:
                print("WTF??")
        elif char == ";":
            FLAG = FLAG_COMMENT
        else:
            buffer.append(char)
    if len(last) != 1 or (FLAG != FLAG_DEFAULT and FLAG != FLAG_COMMENT):
        # print(FLAG)
        raise SyntaxError
    if buffer:
        last[-1].append(parse_value(buffer))
        buffer = []
    return res

File: kotlin/src/me/mnahong2112/kf/Parser.kt
import java.util.*

/*
 * 參照某天跟冰封提起的方法嘗試實現的一個解釋器
 * 用VSCode寫的...鬼知道能不能用(攤手
 */

// parser str to list, kotlin version
/*
def parse_value(buffer):
    if is_quote_by(buffer, '"'):
    return String(parse_string(buffer[1:-1]))
    s = ''.join(buffer)
    if s.startswith("0x"):
    return Number(int(s, 16))
    if s.startswith("0b") and "." not in s:
    return Number(int(s, 2))
    if s.startswith("0o") and "." not in s:
    return Number(int(s, 8))
    try:
    return Number(int(s))
    except Exception: pass
    try:
    return Number(float(s))
    except Exception: pass
    return Symbol(s)
*/

class Reader<T>(val arr: Collection<T>): Iterable<T>, Iterator<T> {
   override fun hasNext(): Boolean {
      return index < length
   }

   override fun iterator(): Iterator<T> {
      return this
   }

   val length = arr.size
   var index = 0

   override fun next(): T {
      index += 1
      if (index < length) {
         return arr.elementAt(index)
      } else {
         throw IndexOutOfBoundsException()
      }
   }

   fun next(n: Int): ArrayList<T> {
      if (index < length) {
         val r = ArrayList<T>(n)
         var k = 0
         for(i in index .. Math.min(index+n, length) - 1) {
            r[k++] = arr.elementAt(i)
         }
         index += n
         index = Math.min(index, length)
         return r
      } else {
         throw IndexOutOfBoundsException()
      }
   }
}

class Parser() {
   sealed class Token {
      data class String(val value: kotlin.String) : Token()
      data class Symbol(val value: kotlin.String) : Token()
      data class Number(val value: kotlin.Number) : Token()
      object OpenBracket : Token()
      object CloseBracket : Token()
   }

   fun valueParser(buffer: StringBuilder): Token {
      if (isQuoteBy(buffer, '"')) {
         return stringParser(buffer.substring(1, buffer.length - 1))
      }
      val s = buffer.toString()
      when (true) {
         s.startsWith("0x") -> {
            return Token.Number(Integer.valueOf(s, 16))
         }
         s.startsWith("0b") -> {
            return Token.Number(Integer.valueOf(s, 2))
         }
         s.startsWith("0o") -> {
            return Token.Number(Integer.valueOf(s, 8))
         }
         s[0].isDigit() -> {
            return Token.Number(Integer.valueOf(s))
         }
         else -> {
            return Token.Symbol(s)
         }
      }
   }

   fun stringParser(s: String): Token.String {

   }

   fun isQuoteBy(buffer: StringBuilder, c: Char): Boolean {
      return buffer.length > 1 && buffer[0] == c && buffer[buffer.length - 1] == c
   }

   val FLAG_DEFAULT = 0
   val FLAG_STRING = 1
   val FLAG_ESCAPING_STRING = 2
   val FLAG_COMMENT = 3

   fun preProcess(expr: String): Stack<Token> {

   }
   fun parse(expr: String): Stack<Token> {
      val res = Expr()
      val last = Stack<Expr>()
      val R = Reader(preProcess(expr))
      for(i in R) {
         when(i) {
            is Token.OpenBracket -> {
               val new = Stack<Token>()
               last.peek().push(new)
               last.push(new)
            }
         }
      }
      if type(obj) is str:
      if obj in "([{":
      new = []
      last[-1].append(new)
      last.append(new)
      elif obj in ")]}":
      l = last.pop()
      else:
      last[-1].append(obj)
      else:
      last[-1].append(obj)
      return res
   }
}

